Setting custom units in `spectrum_from_column_mapping` fails with astropy 4.3

These tests on v1.3 are failing with astropy 4.3rc1:
```python
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_writer[wavelength] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_writer[frequency] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_writer[energy] - KeyError: 'energy'
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_writer[wavenumber] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[wavelength-1] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[wavelength-2] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[wavelength-3] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[frequency-1] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[frequency-2] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[frequency-3] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[energy-1] - KeyError: 'energy'
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[energy-2] - KeyError: 'energy'
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[energy-3] - KeyError: 'energy'
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[wavenumber-1] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[wavenumber-2] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
FAILED lib/python3.9/site-packages/specutils/tests/test_loaders.py::test_tabular_fits_multid[wavenumber-3] - assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
```
the majority are failures on the `flux` unit like
```python
        spectrum.write(tmpfile, format='tabular-fits', overwrite=True)
    
        cmap = {spectral_axis: ('spectral_axis', wlu[spectral_axis]),
                'flux': ('flux', 'erg / (s cm**2 AA)'),
                'uncertainty': ('uncertainty', None)}
    
        # Read it back again and check against the original
        spec = Spectrum1D.read(tmpfile, format='tabular-fits', column_mapping=cmap)
>       assert spec.flux.unit == u.Unit('erg / (s cm**2 AA)')
E       assert Unit("W / (Angstrom m2)") == Unit("erg / (Angstrom cm2 s)")
E        +  where Unit("W / (Angstrom m2)") = <Quantity [ 5.15047686e-10,  3.85273149e-09,  5.70890511e-10,\n            1.13556564e-09,  9.54001763e-10,  6.51391251...     -3.15269245e-10,  7.58969220e-10, -7.72825215e-10,\n           -2.36818607e-10, -4.85363548e-10] W / (Angstrom m2)>.unit
E        +    where <Quantity [ 5.15047686e-10,  3.85273149e-09,  5.70890511e-10,\n            1.13556564e-09,  9.54001763e-10,  6.51391251...     -3.15269245e-10,  7.58969220e-10, -7.72825215e-10,\n           -2.36818607e-10, -4.85363548e-10] W / (Angstrom m2)> = <Spectrum1D(flux=<Quantity [ 5.15047686e-10,  3.85273149e-09,  5.70890511e-10,\n            1.13556564e-09,  9.54001763...                  1.77558228e-06, 2.75493960e-06, 2.77997341e-06,\n                   1.53889118e-06, 2.20309679e-06]))>.flux
E        +  and   Unit("erg / (Angstrom cm2 s)") = <class 'astropy.units.core.Unit'>('erg / (s cm**2 AA)')
E        +    where <class 'astropy.units.core.Unit'> = u.Unit
```
the reason being `spectrum_from_column_mapping` is failing the check
https://github.com/astropy/specutils/blob/3a86739d53663b528ee751eb55d01de3c252022d/specutils/io/parsing_utils.py#L110
as `unit.physical_type` in astropy 4.3 has become an object of type `PhysicalType` instead of a plain string; this can be simply fixed by checking for `'spectral flux' in str(cm_unit.physical_type)` instead.

Somewhat related, `spectral_axis` type 'energy' is not recognised since energy units like `u.eV` have now become
`PhysicalType({'energy', 'torque', 'work'})`
and are accordingly written to FITS as
`TTYPE1  = 'energy/torque/work'`

I have to think about how to proceed best with this case. The `spectral_axis` column would need to be referenced as `table['energy/torque/work']`, but I'd prefer to have the writer not use that ugly `TTYPE` in the first place.
Setting custom units on 'energy' seems broken anyway, as the test simply uses `u.eV` before _and_ after.

--- a/specutils/io/default_loaders/tabular_fits.py
+++ b/specutils/io/default_loaders/tabular_fits.py
@@ -134,10 +134,12 @@ def tabular_fits_writer(spectrum, file_name, hdu=1, update_header=False, **kwarg
     wunit = u.Unit(kwargs.pop('wunit', spectrum.spectral_axis.unit))
     disp = spectrum.spectral_axis.to(wunit, equivalencies=u.spectral())
 
-    # Mapping of spectral_axis types to header TTYPE1
-    dispname = wunit.physical_type
+    # Mapping of spectral_axis types to header TTYPE1 (no "torque/work" types!)
+    dispname = str(wunit.physical_type)
     if dispname == "length":
         dispname = "wavelength"
+    elif "energy" in dispname:
+        dispname = "energy"
 
     # Add flux array and unit
     ftype = kwargs.pop('ftype', spectrum.flux.dtype)

--- a/specutils/io/parsing_utils.py
+++ b/specutils/io/parsing_utils.py
@@ -104,8 +104,8 @@ def spectrum_from_column_mapping(table, column_mapping, wcs=None):
 
             if not isinstance(cm_unit, u.Unit):
                 cm_unit = u.Unit(cm_unit)
-            cm_type =  str(cm_unit.physical_type)
-            if 'length in cm_type' or 'frequency' in cm_type or 'energy' in cm_type:
+            cm_type = str(cm_unit.physical_type)
+            if 'length' in cm_type or 'frequency' in cm_type or 'energy' in cm_type:
                 # Spectral axis column information
                 kwarg_val = kwarg_val.to(cm_unit, equivalencies=u.spectral())
             elif 'spectral flux' in cm_type:

--- a/specutils/io/parsing_utils.py
+++ b/specutils/io/parsing_utils.py
@@ -104,11 +104,10 @@ def spectrum_from_column_mapping(table, column_mapping, wcs=None):
 
             if not isinstance(cm_unit, u.Unit):
                 cm_unit = u.Unit(cm_unit)
-            cm_type = str(cm_unit.physical_type)
-            if 'length' in cm_type or 'frequency' in cm_type or 'energy' in cm_type:
+            if cm_unit.physical_type in ('length', 'frequency', 'energy'):
                 # Spectral axis column information
                 kwarg_val = kwarg_val.to(cm_unit, equivalencies=u.spectral())
-            elif 'spectral flux' in cm_type:
+            elif 'spectral flux' in str(cm_unit.physical_type):
                 # Flux/error column information
                 kwarg_val = kwarg_val.to(cm_unit, equivalencies=u.spectral_density(1 * u.AA))
         elif tab_unit:

--- a/specutils/io/parsing_utils.py
+++ b/specutils/io/parsing_utils.py
@@ -99,8 +99,8 @@ def spectrum_from_column_mapping(table, column_mapping, wcs=None):
             kwarg_val = u.Quantity(table[col_name], tab_unit)
 
             # Attempt to convert the table unit to the user-defined unit.
-            log.debug(f"Attempting auto-convert of table unit {tab_unit} to "
-                      f"user-provided unit {cm_unit}.")
+            log.debug("Attempting auto-convert of table unit '%s' to "
+                      "user-provided unit '%s'.", tab_unit, cm_unit)
 
             if not isinstance(cm_unit, u.Unit):
                 cm_unit = u.Unit(cm_unit)

--- a/specutils/io/parsing_utils.py
+++ b/specutils/io/parsing_utils.py
@@ -99,18 +99,18 @@ def spectrum_from_column_mapping(table, column_mapping, wcs=None):
             kwarg_val = u.Quantity(table[col_name], tab_unit)
 
             # Attempt to convert the table unit to the user-defined unit.
-            log.debug("Attempting auto-convert of table unit '%s' to "
-                          "user-provided unit '%s'.", tab_unit, cm_unit)
+            log.debug(f"Attempting auto-convert of table unit {tab_unit} to "
+                      f"user-provided unit {cm_unit}.")
 
             if not isinstance(cm_unit, u.Unit):
                 cm_unit = u.Unit(cm_unit)
-            if cm_unit.physical_type in ('length', 'frequency'):
+            cm_type =  str(cm_unit.physical_type)
+            if 'length in cm_type' or 'frequency' in cm_type or 'energy' in cm_type:
                 # Spectral axis column information
                 kwarg_val = kwarg_val.to(cm_unit, equivalencies=u.spectral())
-            elif 'spectral flux' in cm_unit.physical_type:
+            elif 'spectral flux' in cm_type:
                 # Flux/error column information
-                kwarg_val = kwarg_val.to(
-                    cm_unit, equivalencies=u.spectral_density(1 * u.AA))
+                kwarg_val = kwarg_val.to(cm_unit, equivalencies=u.spectral_density(1 * u.AA))
         elif tab_unit:
             # The user has provided no unit in the column mapping, so we
             # use the unit as defined in the table object.

--- a/specutils/tests/test_loaders.py
+++ b/specutils/tests/test_loaders.py
@@ -522,14 +522,14 @@ def test_tabular_fits_writer(tmpdir, spectral_axis):
         spectrum.write(tmpfile, format='tabular-fits')
     spectrum.write(tmpfile, format='tabular-fits', overwrite=True)
 
-    cmap = {spectral_axis: ('spectral_axis', wlu[spectral_axis]),
+    cmap = {spectral_axis: ('spectral_axis', 'micron'),
             'flux': ('flux', 'erg / (s cm**2 AA)'),
             'uncertainty': ('uncertainty', None)}
 
     # Read it back again and check against the original
     spec = Spectrum1D.read(tmpfile, format='tabular-fits', column_mapping=cmap)
     assert spec.flux.unit == u.Unit('erg / (s cm**2 AA)')
-    assert spec.spectral_axis.unit == spectrum.spectral_axis.unit
+    assert spec.spectral_axis.unit == u.um
     assert quantity_allclose(spec.spectral_axis, spectrum.spectral_axis)
     assert quantity_allclose(spec.flux, spectrum.flux)
     assert quantity_allclose(spec.uncertainty.quantity,
@@ -566,13 +566,13 @@ def test_tabular_fits_multid(tmpdir, ndim, spectral_axis):
                              spectrum.uncertainty.quantity)
 
     # Test again, using `column_mapping` to convert to different flux unit
-    cmap = {spectral_axis: ('spectral_axis', wlu[spectral_axis]),
+    cmap = {spectral_axis: ('spectral_axis', 'THz'),
             'flux': ('flux', 'erg / (s cm**2 AA)'),
             'uncertainty': ('uncertainty', None)}
 
     spec = Spectrum1D.read(tmpfile, format='tabular-fits', column_mapping=cmap)
     assert spec.flux.unit == u.Unit('erg / (s cm**2 AA)')
-    assert spec.spectral_axis.unit == spectrum.spectral_axis.unit
+    assert spec.spectral_axis.unit == u.THz
     assert quantity_allclose(spec.spectral_axis, spectrum.spectral_axis)
     assert quantity_allclose(spec.flux, spectrum.flux)
     assert quantity_allclose(spec.uncertainty.quantity,

--- a/specutils/tests/test_loaders.py
+++ b/specutils/tests/test_loaders.py
@@ -522,6 +522,7 @@ def test_tabular_fits_writer(tmpdir, spectral_axis):
         spectrum.write(tmpfile, format='tabular-fits')
     spectrum.write(tmpfile, format='tabular-fits', overwrite=True)
 
+    # Map to alternative set of units
     cmap = {spectral_axis: ('spectral_axis', 'micron'),
             'flux': ('flux', 'erg / (s cm**2 AA)'),
             'uncertainty': ('uncertainty', None)}
@@ -565,7 +566,7 @@ def test_tabular_fits_multid(tmpdir, ndim, spectral_axis):
     assert quantity_allclose(spec.uncertainty.quantity,
                              spectrum.uncertainty.quantity)
 
-    # Test again, using `column_mapping` to convert to different flux unit
+    # Test again, using `column_mapping` to convert to different spectral axis and flux units
     cmap = {spectral_axis: ('spectral_axis', 'THz'),
             'flux': ('flux', 'erg / (s cm**2 AA)'),
             'uncertainty': ('uncertainty', None)}

